### PR TITLE
upgrades: enforce specific upgrades for V??_?Start keys

### LINKED ISSUE #104418: upgrades: catalog health checks should run as a precondition to upgrades
We had this in the past in the upgrade path to 22.2 in the form of a tenant upgrade step `preconditionBeforeStartingAnUpgrade` bound to `clusterversion.Start22_2`. Unfortunately there is no equivalent for `clusterversion.Start23_1`. We might want to retrofit this into the 23.1 release, unfortunately it's already gone out, so we'd have to add this check within an existing 23.1 upgrade step.

Remembering to re-add the precondition check for each release `Start*` version is not sustainable. Instead, perhaps we can modify the `upgrademanager.Manager`'s `Migrate` method to run these checks there, or something to that effect.

The check should automatically repair simple cases of corruption if it's known to be safe. For example: https://github.com/cockroachdb/cockroach/issues/85265

Jira issue: CRDB-28520

Epic CRDB-28665

### PR DESCRIPTION
**Pull request by postamar**
Previously, we'd added the precondition check functionality to cluster
upgrade steps in large part because we wanted to check for descriptor
corruptions before actually proceeding with a cluster upgrade. Otherwise
what might (and did) happen was that an unrelated upgrade step would get
stuck and throw an error while the corruption was still present, leaving
the cluster upgrade in limbo in some intermediate internal version.
Descriptor corruption can fly under the radar for a long time in some
cases (like dangling back-references) because these might not otherwise
cause any problems until a DROP is performed.

The problem is that this precondition check was tied to a specific
cluster version key which has since been deleted as part of the regular
release engineering process and there were no tests or any other
mechanism that would ensure that cluster version upgrades would always
include this precondition.

This commit enforces the presence of upgrade steps for all V??_?Start
cluster version keys, which have internal version 2 and succeed each
supported pre-existing release for this binary. These upgrade steps
will be the first to run anytime a cluster version upgrade is performed.
This removes the need from having to remember adding this every time a
major release is done.

Presently, these have a precondition check that verifies that the
invalid_objects vtable is empty. The upgrade step itself consists of
replacing all entries in system.descriptor with protobufs having
undergone post-deserialization changes when applicable. Future work
will extend these metadata health checks and repairs appropriately.

Fixes #104418.

Release note (general change): henceforth, all cluster version upgrades perform a catalog health check and repair upgrade step before proceeding to other upgrade steps.

### CODE REVIEW

**Comment by postamar** [comment]
cc @rafiss @dt @knz, I would appreciate all of you taking a look at this. Do we like having this kind of implicit upgrade step? I think it's OK because it's unlikely and perhaps even undesirable to define specific upgrade steps for `V23_2Start` and friends.

**Review by knz — COMMENTED** [review]
I would welcome more words in the explanation, with a section that first focuses on the problem that existed before your PR (and the impact of that problem) before reading about the solution.

(One thought that sprang to mind when I first looked at the code is -- why not a regular migration, and a separate unit test that asserts that the migration entry is always present? So the test would ensure we don't forget to have it there. But I want to understand the problem first.)

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/105657#-:-NZ0uK4H1joVfx94qkDT:b-25tmiq)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @postamar)

___
*[`pkg/upgrade/upgrades/first_upgrade.go` line 97 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/105657#-NZ0uEpK37wXCjGQhJ9z:-NZ0uEpK37wXCjGQhJA-:bdpodj0) ([raw file](https://github.com/cockroachdb/cockroach/blob/4018be6d615775667212cef44f680d2b19c75dbd/pkg/upgrade/upgrades/first_upgrade.go#L97)):*
> ```Go
> 	}
> 	if n := row[0].String(); n != "0" {
> 		return errors.Errorf("%s row(s) found in \"\".crdb_internal.invalid_objects, expected none", n)
> ```

nit: `errors.AssertionFailedf`

**Comment by postamar** [comment]
Thanks for taking a look @knz 

> I would welcome more words in the explanation, with a section that first focuses on the problem that existed before your PR (and the impact of that problem) before reading about the solution.

Will do.

> (One thought that sprang to mind when I first looked at the code is -- why not a regular migration, and a separate unit test that asserts that the migration entry is always present? So the test would ensure we don't forget to have it there. But I want to understand the problem first.)

This is, in fact, the path I went down initially. The challenge was what to assert on to check that the upgrade precondition and run functions are what we expect them to be. These are private to the parent package and I didn't feel it was reasonable to expose methods on TenantUpgrade or introduce a subtype just for the purpose of this test. I found it less brittle and more straightforward to define the upgrade step constructively in the init function. It's IMO also very low risk, I can't think of a situation where anyone would deliberately want to bind an upgrade to a V??_?Start cluster version key, instead of simply adding a new key 2 internal versions beyond.

> nit: errors.AssertionFailedf

Is it, though? I wondered about this. You're probably right, though. Will do.

**Comment by dt** [comment]
> These are private to the parent package and I didn't feel it was reasonable to expose methods on TenantUpgrade or introduce a subtype just for the purpose of this test.

Maybe I'm missing something: Can you not just do `var validateCatalogUpgrade = upgrade.NewTenantUpgrade(.... validationFunc...)` in `validate_catalog_upgrade.go` and then add ` have a unit test in pkg/upgrade/upgrades that looks for, using simple pointer equality, `validateCatalogUpgrade` in the `upgrades` slice (potentially specifically at index 2)?

**Comment by knz** [comment]
> I remain unconvinced of the value this approach would add

What I am concerned about is the cognitive overhead to folk who are looking at the "production code path" to understand how it works and what they need to care about. Your PR is adding a new _abstraction_ which wasn't before so it is really extra conceptual complexity. Fitting the mechanism under the existing abstraction and adding a test, in contrast, doesn't introduce extra conceptual complexity.

**Comment by postamar** [comment]
Meta-thought: our abstractions for reasoning about versions are too weak. They always were annoying to reason about but the need to cover more than one pre-existing release has upped this annoyance to a whole other level. Why can't I easily get an ordered collection of keys for pre-existing major releases? Why do I always have to trip on off-by-one errors when reasoning about internal versions, roachpb.Version's need to be monotonically increasing conflicts with the intuition that an upgrade step "belongs" to the _target_ release version rather than the _origin_. I don't have any ideas for cheap and easy improvements (less than a week's worth of work) but if any of you do please let me know.

**Comment by dt** [comment]
>  I'd also have to check that all releases are covered
The release team has a runbook they follow pretty diligently for major versions; if we add "add the validateCatalog upgrade after VX_Start` to their runbook and code comments, they will do it.

**Comment by postamar** [comment]
RFAL

**Review by knz — APPROVED** [review]
> Reviewed 1 of 4 files at r1, 2 of 3 files at r2, 3 of 3 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/105657)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @postamar)

___
*[`pkg/upgrade/upgrades/first_upgrade.go` line 97 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/105657#-NZ0uEpK37wXCjGQhJ9z:-NZ1gvBo3LqBeSici93l:b-r9uksk) ([raw file](https://github.com/cockroachdb/cockroach/blob/4018be6d615775667212cef44f680d2b19c75dbd/pkg/upgrade/upgrades/first_upgrade.go#L97)):*
<details><summary><i>Previously, knz (Raphael 'kena' Poss) wrote…</i></summary><blockquote>

nit: `errors.AssertionFailedf`
</blockquote></details>

Beyond the basic framework, I'd like to motion for extra code here that would push a sentry report out.
What do you think of a call to `logcrash.ReportOrPanic`?

**Comment by postamar** [comment]
Thanks for the review. Wouldn't wrapping this as an assertion failure be enough to trigger a sentry report, though? I thought those went hand in hand. Perhaps I'm missing something.

**Comment by knz** [comment]
> Wouldn't wrapping this as an assertion failure be enough to trigger a sentry report, though?

The decision about whether or not to send a sentry report is not made where the error is constructed. It's made at some specific API boundaries.

For example, we have such a boundary at the SQL network interface. Assertions that make their way to the client will also get copied to Sentry. We have other such boundaries.

Here the error is definitely not going to the SQL client boundary. I wonder where else it goes though.
- interactive uses of `set cluster setting version = ...` do flow the migration error back to the client so that is covered
- but we also have the code in `server/auto_upgrade.go` where it's merely logged.

I wonder if we want to recognize assertion failures in the logging package? We don't do that yet.

(Also -- all this conversation is orthogonal to your PR and is non-blocking. Feel free to move it to a separate issue.)

**Comment by postamar** [comment]
Got it, thanks. This probably should be done for all preconditions and upgrade errors, not just this one. I'll file an issue.

bors r+

File: pkg/upgrade/upgrades/upgrades.go [hunk 1]
```diff
@@ -318,11 +319,34 @@ var upgrades = []upgradebase.Upgrade{
...
+		firstUpgradeTowardsV23_1,
+		firstUpgradeTowardsV23_2,
+	}
+)
+
 func init() {
 	for _, m := range upgrades {
-		if _, exists := registry[m.Version()]; exists {
-			panic(errors.AssertionFailedf("duplicate upgrade registration for %v", m.Version()))
-		}
```
Comment (Xiang-Gu) [inline pkg/upgrade/upgrades/upgrades.go:325]: why do we remove this block?

Comment (Xiang-Gu) [reply pkg/upgrade/upgrades/upgrades.go:325]: nvm, I saw the TestUniqueVersions test below for this

**Review by Xiang-Gu — COMMENTED** [review]
I have a question "How did we, in a unit test, ensure we don't forget to set the Precondition and UpgradeFunc for future major releases?" 
I don't think the test currently would complain if we added a new `firstUpgradeTowardsV24_1` with a nil Precondition and nil UpgradeFunc.

**Comment by postamar** [comment]
You're right, it relies on the release engineers not making mistakes.

**Pull request merged**